File: solutions/clojure/poker/1/src/poker.clj
(ns poker
  (:require [clojure.string :as st]))


(defn ->nbr [s]
  (read-string
   (st/replace s
               #"[AJQK]"
               {"A" "14" "J" "11" "Q" "12" "K" "13"})))


(defn by-suit [cards]
  (->> cards
       (group-by second)
       (map (fn [[k cards]] [k (sort (map first cards))]))
       (into {})))


(defn by-rank [cards]
  (->> cards
       (map first)
       frequencies))


(defn ->cards [hand]
  (->> hand
       (re-seq #"([\dAJQK]+)([SHDC])")
       (map rest)
       (map (fn [[card suit]] [(->nbr card) suit]))
       ((juxt by-suit by-rank))
       (zipmap [:suits :cards])))

(defn same-suit? [n [_ cards]]
  (= n (count cards)))

(defn in-range? [nbrs]
  (->> nbrs
       (map - (rest nbrs))
       (every? (partial = 1))))


(defn straight-flush [{:keys [suits]}]
  (->> suits
       (filter (partial same-suit? 5))
       first
       second
       (#(when (and % (in-range? %))
           (apply + 900 %)))))


(defn four-of-kind
  "4 cards of the same rank"
  [{:keys [cards]}]
  (let [[[n1 t1] [n2 t2] third] (sort-by second cards)]
    (and (nil? third)
         (= 4 t2)
         (+ 800 (* n2 t2) (/ (* t1 n1) 100.0)))))


(defn full-house
  "3 cards of one rank and 2 cards of another rank"
  [{:keys [cards]}]
  (let [[[n1 t1] [n2 t2] third] (sort-by second cards)]
    (and (nil? third)
         (= 2 t1)
         (+ 700 (* n1 t1) (* n2 t2)))))


(defn sflush [{:keys [suits]}]
  (->> suits
       (filter (partial same-suit? 5))
       first
       second
       (#(when % (apply + 600 %)))))

(def straights?
  (conj
   (set
    (partition 5
               1
               [2 3 4 5 6 7 8 9 10 11 12 13 14]))
   [2 3 4 5 14]))


(defn straight [{:keys [cards]}]
  (when-let [sts (straights? (sort (keys cards)))]
    (let [ks (if (= [2 3 4 5 14] sts) (range 6) sts)]
      (apply + 500 ks))))


(defn three-of-kind
  "3 cards of the same rank"
  [{:keys [cards]}]
  (let [[[n1 t1] [n2 t2] [n3 t3]] (sort-by second > cards)]
    (and (= 3 t1)
         (= 1 t2)
         (+ 400 n1 (/ (+ (* n2 t2) (* n3 t3)) 100.0)))))

(defn two-pair [{:keys [cards]}]
  (let [[[n1 t1] [n2 t2] [n3 t3]] (sort-by second > cards)]
    (and (= 2 t1 t2)
         (+ 300
            n1
            n2
            (/ (* n3 t3) 100.0)))))

(defn one-pair [{:keys [cards]}]
  (let [[[n1 t1]] (sort-by second > cards)]
    (and (= 2 t1)
         (+ 200
            n1))))


(defn high-card [{:keys [cards]}]
  (->> cards
       keys
       sort
       (reduce #(+ %2 (/ %1 10.0)))))


(defn score [hand]
  (some #(% hand)
        [straight-flush
         four-of-kind
         full-house
         sflush
         straight
         three-of-kind
         two-pair
         one-pair
         high-card]))

(defn highest-hands [[[_ h] :as hands]]
  (->> hands
       (filter (comp (partial = h) second))
       (map first)))

(defn best-hands [hands]
  (->> hands
       (map (juxt identity (comp score ->cards)))
       (sort-by second >=)
       highest-hands))

File: solutions/python/poker/1/poker.py
"""Choose the best poker hands from a collection of five-card hands."""

from collections import Counter


VALUES = {str(value): value for value in range(2, 11)} | {"J": 11, "Q": 12, "K": 13, "A": 14}


def _hand_key(hand: str) -> tuple:
    """Return a comparable category and tie-break key for a hand."""
    cards = [(VALUES[card[:-1]], card[-1]) for card in hand.split()]
    values = sorted((value for value, _ in cards), reverse=True)
    unique = set(values)

    counts = Counter(values)
    groups = sorted(((count, value) for value, count in counts.items()), reverse=True)
    flush = len({suit for _, suit in cards}) == 1
    straight_high = (
        5
        if unique == {14, 2, 3, 4, 5}
        else max(unique)
        if len(unique) == 5 and max(unique) - min(unique) == 4
        else 0
    )

    if flush and straight_high:
        category, tiebreak = 8, (straight_high,)
    elif groups[0][0] == 4:
        category, tiebreak = 7, (groups[0][1], groups[1][1])
    elif groups[0][0] == 3 and groups[1][0] == 2:
        category, tiebreak = 6, (groups[0][1], groups[1][1])
    elif flush:
        category, tiebreak = 5, tuple(values)
    elif straight_high:
        category, tiebreak = 4, (straight_high,)
    elif groups[0][0] == 3:
        kickers = sorted((value for value in values if value != groups[0][1]), reverse=True)
        category, tiebreak = 3, (groups[0][1], *kickers)
    elif groups[0][0] == 2 and groups[1][0] == 2:
        pairs = sorted((groups[0][1], groups[1][1]), reverse=True)
        kicker = groups[2][1]
        category, tiebreak = 2, (*pairs, kicker)
    elif groups[0][0] == 2:
        pair = groups[0][1]
        kickers = sorted((value for value in values if value != pair), reverse=True)
        category, tiebreak = 1, (pair, *kickers)
    else:
        category, tiebreak = 0, tuple(values)

    return (category, *tiebreak)


def best_hands(hands):
    """Return all hands tied for the highest poker ranking."""
    scored_hands = [(hand, _hand_key(hand)) for hand in hands]
    best_score = max(score for _, score in scored_hands)
    return [hand for hand, score in scored_hands if score == best_score]
